File: app/src/main/java/com/example/compass/libs/utils/ConversionUtils.java
package com.example.compass.libs.utils;

public final class ConversionUtils {

    private ConversionUtils() {

    }

    public static void convertInchFractions(float value, int[] fractions) {

        // denominator is fixed
        int denominator = 32;
        // integer part, can be signed: 1, 0, -3,...
        int integer = (int) value;
        // numerator: always unsigned (the sign belongs to the integer part)
        // + 0.5 - rounding, nearest one: 37.9 / 64 -> 38 / 64; 38.01 / 64 -> 38 / 64
        int numerator = (int) ((Math.abs(value) - Math.abs(integer)) * denominator + 0.5);

        // some fractions, e.g. 24 / 64 can be simplified:
        // both numerator and denominator can be divided by the same number
        // since 64 = 2 ** 6 we can try 2 powers only
        // 24/64 -> 12/32 -> 6/16 -> 3/8
        // In general case (arbitrary denominator) use gcd (Greatest Common Divisor):
        // double factor = gcd(denominator, numerator);
        // denominator /= factor;
        // numerator /= factor;
        while ((numerator % 2 == 0) && (denominator % 2 == 0)) {
            numerator /= 2;
            denominator /= 2;
        }

        fractions[2] = denominator;
        fractions[1] = numerator;
        fractions[0] = (int) integer + numerator / denominator;
    }


    /**
     * Get rounded float value
     *
     * @param digit  represents decimal places
     * @param number to be rounded
     * @return rounded float number
     */
    public static float getDecimalPlaceFloat(int digit, float number) {

        switch (digit) {
            case 0:
                number = Math.round(number);
                break;
            case 1:
                number = Math.round(number * 10f) / 10f;
                break;

            case 2:
                number = Math.round(number * 100f) / 100f;
                break;
        }
        return number;
    }

    public static float convertDegreesToPercent(float degrees) {
        float percent = (float) Math.tan(degrees * Math.PI / 180);
        if (!Float.isNaN(percent)) {
            percent = percent * 100;
            return percent;
        } else {
            return 1000;
        }
    }

    public static float convertPercentToDegrees(float percent) {
        float degrees = (float) (Math.atan(percent / 100) * 180 / Math.PI);
        if (!Float.isNaN(degrees)) {
            return degrees;
        } else {
            return 90;
        }
    }
}

File: app/src/main/java/com/smarttoolfactory/compass/libs/utils/ConversionUtils.kt
package com.smarttoolfactory.compass.libs.utils

import kotlin.math.abs
import kotlin.math.atan
import kotlin.math.roundToInt
import kotlin.math.tan

object ConversionUtils {

    fun convertInchFractions(value: Float, fractions: IntArray) {

        // denominator is fixed
        var denominator = 32
        // integer part, can be signed: 1, 0, -3,...
        val integer = value.toInt()
        // numerator: always unsigned (the sign belongs to the integer part)
        // + 0.5 - rounding, nearest one: 37.9 / 64 -> 38 / 64; 38.01 / 64 -> 38 / 64
        var numerator = ((abs(value) - abs(integer)) * denominator + 0.5).toInt()

        // some fractions, e.g. 24 / 64 can be simplified:
        // both numerator and denominator can be divided by the same number
        // since 64 = 2 ** 6 we can try 2 powers only
        // 24/64 -> 12/32 -> 6/16 -> 3/8
        // In general case (arbitrary denominator) use gcd (Greatest Common Divisor):
        // double factor = gcd(denominator, numerator);
        // denominator /= factor;
        // numerator /= factor;
        while (numerator % 2 == 0 && denominator % 2 == 0) {
            numerator /= 2
            denominator /= 2
        }
        fractions[2] = denominator
        fractions[1] = numerator
        fractions[0] = integer + numerator / denominator
    }

    /**
     * Get rounded float value
     *
     * @param digit  represents decimal places
     * @param number to be rounded
     * @return rounded float number
     */
    fun getDecimalPlaceFloat(digit: Int, number: Float): Float {
        var number = number
        when (digit) {
            0 -> number = number.roundToInt().toFloat()
            1 -> number = (number * 10f).roundToInt() / 10f
            2 -> number = (number * 100f).roundToInt() / 100f
        }
        return number
    }

    fun convertDegreesToPercent(degrees: Float): Float {
        var percent = tan(degrees * Math.PI / 180).toFloat()
        return if (!java.lang.Float.isNaN(percent)) {
            percent *= 100
            percent
        } else {
            1000f
        }
    }

    fun convertPercentToDegrees(percent: Float): Float {
        val degrees = (atan(percent / 100.toDouble()) * 180 / Math.PI).toFloat()
        return if (!java.lang.Float.isNaN(degrees)) {
            degrees
        } else {
            90f
        }
    }
}